openmc/data/resonance_covariance.py

```diff
@@ -0,0 +1,812 @@
+                    energy = sample[0::5]
+                    gn = sample[1::5]
+                    gg = sample[2::5]
+                    gfa = sample[3::5]
+                    gfb = sample[4::5]
+                    records = []
+                    for j, E in enumerate(energy):
+                        records.append([energy[j], l_value[j], spin[j], gn[j],
+                                        gg[j], gfa[j], gfb[j]])
+                    columns = ['energy', 'L', 'J', 'neutronWidth',
+                               'captureWidth', 'fissionWidthA', 'fissionWidthB']
+                    sample_params = pd.DataFrame.from_records(records, columns=columns)
+                    res_range = copy.copy(resonances)
+                    res_range._prepared = False # Set prepared to False to ensure
+                                                # the sampled parameters are used
+                                                # in reconstruction
+                    res_range.parameters = sample_params
+                    samples.append(res_range)
+    
+        self.samples = samples
+
+
+class MultiLevelBreitWignerCovariance(ResonanceCovarianceRange):
+    """Multi-level Breit-Wigner resolved resonance formalism covariance data.
+    Parameters
+    ----------
+    energy_min : float
+        Minimum energy of the resolved resonance range in eV
+    energy_max : float
+        Maximum energy of the resolved resonance range in eV
+
+    Attributes
+    ----------
+    energy_min : float
+        Minimum energy of the resolved resonance range in eV
+    energy_max : float
+        Maximum energy of the resolved resonance range in eV
+    parameters : pandas.DataFrame
+        Resonance parameters
+    covariance : numpy.array
+        The covariance matrix contained within the ENDF evaluation
+    lcomp : int
+        Flag indicating the format of the covariance matrix within the ENDF file
+    mpar : int
+        Number of parameters in covariance matrix for each individual resonance
+    formalism : str
+        String descriptor of formalism
+    """
+
+    def __init__(self, energy_min, energy_max, parameters, covariance, mpar, lcomp):
+        super().__init__(energy_min, energy_max)
+        self.parameters = parameters
+        self.covariance = covariance
+        self.mpar = mpar
+        self.lcomp = lcomp
+        self.formalism = 'mlbw'
+            
+    @classmethod
+    def from_endf(cls, ev, file_obj, items, file2params):
+        """Create MLBW covariance data from an ENDF evaluation.
+
+        Parameters
+        ----------
+        ev : openmc.data.endf.Evaluation
+            ENDF evaluation
+        file_obj : file-like object
+            ENDF file positioned at the second record of a resonance range
+            subsection in MF=32, MT=151
+        items : list
+            Items from the CONT record at the start of the resonance range
+            subsection
+        file2params : openmc.data.ResonanceRange object
```

Comment:
It looks like this is still actually taking a DataFrame of resonance parameters rather than the `ResonanceRange` object that owns it. I think it would be better to use the `ResonanceRange` object directly and store it as an attribute of the `ResonanceCovarianceRange` instance. That way, when `sample_resonance_parameters` is called, there would be no need to pass the parameters as they could be retrieved from the attribute.

Reply:
Same comment applies to `ReichMooreCovariance`